File: starrysky/src/main/java/com/lzx/starrysky/notification/StarrySkyNotificationManager.java
package com.lzx.starrysky.notification;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.lzx.starrysky.MusicService;
import com.lzx.starrysky.StarrySky;

public class StarrySkyNotificationManager {

    private INotification notification;
    private NotificationFactory factory;
    private boolean isOpenNotification;

    public StarrySkyNotificationManager(boolean isOpenNotification, @Nullable NotificationFactory factory) {
        this.isOpenNotification = isOpenNotification;
        if (isOpenNotification) {
            this.factory = factory == null ? SYSTEM_NOTIFICATION_FACTORY : factory;
        }
    }

    public boolean isOpenNotification() {
        return isOpenNotification;
    }

    public INotification getNotification(MusicService musicService) {
        if (notification == null) {
            synchronized (this) {
                if (notification == null && factory != null) {
                    NotificationConfig config = StarrySky.get().getRegistry().getNotificationConfig();
                    notification = factory.build(musicService, config);
                }
            }
        }
        return notification;
    }


    public interface NotificationFactory {
        @NonNull
        INotification build(MusicService musicService, NotificationConfig config);
    }

    private static final NotificationFactory SYSTEM_NOTIFICATION_FACTORY = SystemNotification::new;

    public static final NotificationFactory CUSTOM_NOTIFICATION_FACTORY = CustomNotification::new;

}

File: starrysky/src/main/java/com/lzx/starrysky/notification/StarrySkyNotificationManager.kt
package com.lzx.starrysky.notification

import android.content.Context

class StarrySkyNotificationManager constructor(
    isOpenNotification: Boolean,
    private val config: NotificationConfig?,
    factory: NotificationFactory?
) {
    private var isOpenNotification: Boolean = false
    private var factory: NotificationFactory? = null
    private var notification: INotification? = null

    init {
        this.isOpenNotification = isOpenNotification
        if (isOpenNotification) {
            this.factory = factory ?: SYSTEM_NOTIFICATION_FACTORY
        }
    }

    fun isOpenNotification(): Boolean {
        return isOpenNotification
    }

    fun getNotification(context: Context): INotification? {
        return notification ?: synchronized(this) {
            if (notification == null && factory != null) {
                notification = factory?.build(context, config)
            }
            return notification
        }
    }

    interface NotificationFactory {
        fun build(context: Context, config: NotificationConfig?): INotification
    }

    companion object {
        val SYSTEM_NOTIFICATION_FACTORY: NotificationFactory = object : NotificationFactory {
            override fun build(
                context: Context, config: NotificationConfig?
            ): INotification {
                return if (config == null) SystemNotification(context) else SystemNotification(context, config)
            }
        }

        val CUSTOM_NOTIFICATION_FACTORY: NotificationFactory = object : NotificationFactory {
            override fun build(
                context: Context, config: NotificationConfig?
            ): INotification {
                return if (config == null) CustomNotification(context) else CustomNotification(context, config)
            }
        }
    }
}